### PR TITLE
Add experiment about integration accuracy.

### PR DESCRIPTION
**Pull request by maxdanwitz**
Inspired by "A toolbox for digital twins" (Mark Asch) Section 3.7.1. Monte Carlo Methods.

### CODE REVIEW

**Comment by isteinbrecher** [comment]
Looks good to me!

File: integration/integrationAccuracyExperiment.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+# As computational domain, we consider a unit square and need two coordinates
+# to parametrize it.
+x, y = symbols("x y")
+
+# This is the function we want to investigate. It is in fact a solution to a
+# Poisson problem taken from ngsolve
+# https://docu.ngsolve.org/latest/whetting_the_appetite/poisson.html
+# Still, let's pretend it is the error of our numerical scheme.
+u = 16 * x * (1 - x) * y * (1 - y)
+# We lamdify it to allow for fast evaluation.
```
Comment (tsarikahin) [inline integration/integrationAccuracyExperiment.py]: lambdify*

Comment (mayrmt) [reply integration/integrationAccuracyExperiment.py]: What do you mean by "lambdifying"?

File: integration/integrationAccuracyExperiment.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+@author: maxvondanwitz
+"""
+import numpy as np
+from sympy import symbols, integrate, sqrt
+from sympy.plotting import plot3d
+from sympy import lambdify
+
+
+def monteCarloIntegration(nn, runs):
+    from numpy import linalg as la
```
Comment (tsarikahin) [inline integration/integrationAccuracyExperiment.py]: I would not do importing libraries in a function, instead, I import them at the beginning always. But as you wish.

File: integration/integrationAccuracyExperiment.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    import random
+
+    res = np.zeros(runs)
```
Comment (tsarikahin) [inline integration/integrationAccuracyExperiment.py:20]: I think this is unnecessary. You can just say

res = la.norm(np.array(vals, dtype=float)) / nn

File: integration/meshBasedInteg.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# Taken from https://docu.ngsolve.org/v6.2.1808/whetting_the_appetite/poisson.html
+
+from ngsolve import *
+from netgen.geom2d import unit_square
+
+# generate a triangular mesh of mesh-size 0.2
+mesh = Mesh(unit_square.GenerateMesh(maxh=0.2))
+
+# to see the really coars mesh, I ran `netgen meshBasedInteg.py` in my shell
```
Comment (BishrMaradni) [inline integration/meshBasedInteg.py]: coarse*

File: integration/meshBasedInteg.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+Draw(mesh)
+
+exact = 16 * x * (1 - x) * y * (1 - y)
+L2 = sqrt(Integrate(exact * exact, mesh))
+print("L2 norm with mesh based integral:", L2)
+print("Relative error:", (8 / 15 - L2) / (8 / 15))
```
Comment (BishrMaradni) [inline integration/meshBasedInteg.py]: 8/15 is the l2 from the analytical solution? How is this reproducible?

File: integration/integrationAccuracyExperiment.py (hunk 1 above)
Comment (BishrMaradni) [reply integration/integrationAccuracyExperiment.py]: From [Sympy.org](https://docs.sympy.org/latest/modules/utilities/lambdify.html):
Transform SymPy expressions to lambda functions which can be used to calculate numerical values very fast.

File: integration/integrationAccuracyExperiment.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        "equidistant sampling points.",
+    )
+
+    print("Relative error:", relativeError(L2, l2))
+# Here, we see O(N^-2). Much better. How to obain (approximately) equidistant
+# samples for more complex geometries? Meshing.
+
+### Option 3: Mesh-based integration
+# works quite well, e.g., using ngsolve.
+l2 = 0.5333333327319278
```
Comment (BishrMaradni) [inline integration/integrationAccuracyExperiment.py:139]: is this l2 obtained from NetGen? Some clarification required

File: integration/integrationAccuracyExperiment.py (hunk 1 above)
Comment (maxdanwitz) [reply integration/integrationAccuracyExperiment.py]: Explanation added as comment.

File: integration/integrationAccuracyExperiment.py (hunk 2 above)
Comment (maxdanwitz) [reply integration/integrationAccuracyExperiment.py]: Thanks for the hint. I adjusted the code accordingly.

File: integration/integrationAccuracyExperiment.py (hunk 6 above)
Comment (maxdanwitz) [reply integration/integrationAccuracyExperiment.py:139]: Clarified as requested.

File: integration/meshBasedInteg.py (hunk 4 above)
Comment (maxdanwitz) [reply integration/meshBasedInteg.py]: Thanks

File: integration/meshBasedInteg.py (hunk 5 above)
Comment (maxdanwitz) [reply integration/meshBasedInteg.py]: Explanation added.

**Review by BishrMaradni — APPROVED** [review]

**Pull request merged**